### PR TITLE
[FLINK-30493][table-api] Introduce TableChange to SET/RESET options

### PR DESCRIPTION
**Pull request by fsk119**
## What is the purpose of the change

*Introduce TableChange to represent ALTER TABLE SET/RESET syntax*

## Brief change log

  - *Introduce TableChange API*
  - *Add Catalog method to alter table with table changes.*
  - *Use AlterTableChangeOperation to represents set/reset*

## Verifying this change

Add change in SqlToOperationConverter to make sure the Operation contains the expected changes.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by fsk119** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 1]
```diff
@@ -595,10 +596,16 @@ private Operation convertAlterTableOptions(
                     new CatalogPartitionImpl(newProps, catalogPartition.getComment()));
         } else {
             // it's altering a table
-            Map<String, String> newOptions = new HashMap<>(oldTable.getOptions());
-            newOptions.putAll(
-                    OperationConverterUtils.extractProperties(alterTableOptions.getPropertyList()));
-            return new AlterTableOptionsOperation(tableIdentifier, oldTable.copy(newOptions));
+            Map<String, String> newOptions =
```
Comment (LadyForest) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:599]: Nit: the naming of `newOptions` and `allOptions` is a little confusing. What about `changeOptions` and `newOptions`?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/TableChange.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    /**
+     * A table change to set the table option.
+     *
+     * <p>It is equal to the following statement:
+     *
+     * <pre>
+     *    ALTER TABLE &lt;table_name&gt; SET '&lt;key&gt;' = '&lt;value&gt;';
+     * </pre>
+     */
+    class SetOption implements TableChange {
```
Comment (LadyForest) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/TableChange.java:75]: Add `@PublicEvolving` annotation?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/TableChange.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    /**
+     * A table change to reset the table option.
+     *
+     * <p>It is equal to the following statement:
+     *
+     * <pre>
+     *    ALTER TABLE &lt;table_name&gt; RESET '&lt;key&gt;'
+     * </pre>
+     */
+    class ResetOption implements TableChange {
```
Comment (LadyForest) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/TableChange.java:127]: ditto

**Comment by LadyForest** [comment]
Hi @fsk119, thanks for the contribution; it looks good in general. I have one question about `AlterTableOptionsOption`. Is it deprecated now? I saw`HiveParserDDLSemanticAnalyzer#convertAlterTableProps` still refer to it.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/TableChange.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        @Override
+        public int hashCode() {
+            return Objects.hash(key);
+        }
+
+        @Override
+        public String toString() {
+            return "ResetOption{" + "key='" + key + '\'' + '}';
+        }
+    }
```
Comment (LadyForest) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/TableChange.java:163]: Nit: Is it mandatory to add `toString`, `equals`, and `hashCode`?

**Comment by fsk119** [comment]
> Is it deprecated now? 

Yes. I think the `AlterTableOptionsOperation ` is deprecated. Considering it's totally internal, we may remove this directly? WDYT?

> I sawHiveParserDDLSemanticAnalyzer#convertAlterTableProps still refer to it.

Thanks for pointing it out, I will add a commit to remove the usage.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/TableChange.java (hunk 4 above)
Comment (fsk119) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/TableChange.java:163]: I think TableChange is a pojo class here. So I add these helpful methods for use.

**Review by LadyForest — APPROVED** [review]
Thanks for the udpate and LGTM :)

**Pull request merged**